src/spawn.rs

```diff
@@ -0,0 +1,463 @@
+            libc::posix_spawn_file_actions_init(
+                &mut self.fa as *mut libc::posix_spawn_file_actions_t,
+            )
+        };
+        if res != 0 {
+            return Err(Errno::from_raw(res));
+        }
+
+        Ok(self)
+    }
+
+    /// Add a [dup2](https://pubs.opengroup.org/onlinepubs/9699919799/functions/dup2.html) action. See
+    /// [posix_spawn_file_actions_adddup2](https://pubs.opengroup.org/onlinepubs/9699919799/functions/posix_spawn_file_actions_adddup2.html).
+    #[doc(alias("posix_spawn_file_actions_adddup2"))]
+    pub fn add_dup2<Fd1: AsFd, Fd2: AsFd>(
+        &mut self,
+        fd: Fd1,
+        newfd: Fd2,
+    ) -> Result<()> {
+        let res = unsafe {
+            libc::posix_spawn_file_actions_adddup2(
+                &mut self.fa as *mut libc::posix_spawn_file_actions_t,
+                fd.as_fd().as_raw_fd(),
+                newfd.as_fd().as_raw_fd(),
+            )
+        };
+        if res != 0 {
+            return Err(Errno::from_raw(res));
+        }
+
+        Ok(())
+    }
+
+    feature! {
+    #![all(feature = "fs", feature = "term")]
+    /// Add an open action. See
+    /// [posix_spawn_file_actions_addopen](https://pubs.opengroup.org/onlinepubs/9699919799/functions/posix_spawn_file_actions_addopen.html).
+    #[doc(alias("posix_spawn_file_actions_addopen"))]
+    pub fn add_open<Fd: AsFd, P: ?Sized + NixPath>(
+        &mut self,
+        fd: Fd,
+        path: &P,
+        oflag: OFlag,
+        mode: Mode,
+    ) -> Result<()> {
+        let res = path.with_nix_path(|cstr| unsafe {
+            libc::posix_spawn_file_actions_addopen(
+                &mut self.fa as *mut libc::posix_spawn_file_actions_t,
+                fd.as_fd().as_raw_fd(),
+                cstr.as_ptr(),
+                oflag.bits(),
+                mode.bits(),
+            )
+        })?;
+        if res != 0 {
+            return Err(Errno::from_raw(res));
+        }
+
+        Ok(())
+    }
+    }
+
+    /// Add a close action. See
+    /// [posix_spawn_file_actions_addclose](https://pubs.opengroup.org/onlinepubs/9699919799/functions/posix_spawn_file_actions_addclose.html).
+    #[doc(alias("posix_spawn_file_actions_addclose"))]
+    pub fn add_close<Fd: AsFd>(&mut self, fd: Fd) -> Result<()> {
+        let res = unsafe {
+            libc::posix_spawn_file_actions_addclose(
+                &mut self.fa as *mut libc::posix_spawn_file_actions_t,
+                fd.as_fd().as_raw_fd(),
+            )
+        };
+        if res != 0 {
+            return Err(Errno::from_raw(res));
+        }
+
+        Ok(())
+    }
+}
+
+impl Drop for PosixSpawnFileActions {
+    fn drop(&mut self) {
+        unsafe {
+            libc::posix_spawn_file_actions_destroy(
+                &mut self.fa as *mut libc::posix_spawn_file_actions_t,
+            );
+        }
+    }
+}
+
+// Specifically for use with posix_spawn and posix_spawnp.
+// https://github.com/rust-lang/libc/issues/1272
```

Comment:
IIUC, the interfaces exposed by the libc crate are actually correct (for posix_spawn and posix_spawnp), it is the POSIX standard's interfaces that are erroneous, so I think we do not need to link to that issue, simply documenting it like this would work:

```
// The POSIX standard requires those `args` and `envp` to be of type `*const *mut [c_char]`, 
// but implementations won't modify them, making the `mut` type redundant. Considering this,
// Nix does not expose this mutability, but we have to change the interface when calling the 
// underlying libc interfaces , this helper function does the conversion job.
//
// SAFETY:
// It is safe to add the mutability in types as implementations won't mutable them.
```

And we can remove the `unsafe` keyword

Reply:
> "implementations won't mutable them" I think you mean mutate here?

Ahh, right, sorry for the typo:)

------

> And we can remove the unsafe keyword

```rs
// SAFETY:
// It is safe to add the mutability in types as implementations won't mutable them.
```

Just realized tha we should not make this function safe but document the `SAFETY` section in its usage